### ISSUE TITLE
Animation skips state transition triggered by RouterOutlet deactivate=>activate

### ISSUE THREAD

**Issue by 4javier** [angular/angular#47284]
### Which @angular/* package(s) are the source of the bug?

animations, router

### Is this a regression?

No

### Description

Not sure if a feature request or a bug.
If you change the value of the property bound to an animation trigger using `activate` and `deactivate` `RouterOutlet`'s outputs, on route navigations where `deactivate` is immediately followed by an `activate` event, the first state change `whatever=>deactivated` gets ignored by animation.
I suppose it's due to the async nature of the trigger. 
In fact turning the outputs property assignments into async makes things work as expected (at least by me).

You can see it in the provided stackblitz

1.  If you let uncommented the sync router-outlet implementation:

-   at first routing you get the correct transition recognized
```
Property changed to: rendered

State transitioned: cleared->rendered
```
- at exiting routing you get correct transition again
```
Property changed to: cleared

State transitioned: rendered->cleared
```
- but at routing between children, when `activated` follows `deactivated`, transitions gets ignored, because seen as `rendered=>rendered`, skipping the middle `cleared`
```
Property changed to: cleared

Property changed to: rendered
```
2. Things works as they should (IMHO) if you comment out the sync router-outlet, and use the one leveraging async assignment on outputs:
this is what you get on routing between children
```
Property changed to: cleared

State transitioned: rendered->cleared

Property changed to: rendered

State transitioned: cleared->rendered
```
I find always "lame" having to resort to setTimeout and similar tricks to turn async something that is not, working in a reactive framework.
So I was wondering if those outputs implementation should be reconsidered.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-ts-strict-enabled-nudgw7?file=src/app/app.component.html

**Comment by dario-piotrowicz** [comment]
Hi @4javier :slightly_smiling_face:, thanks a lot for the very comprehensive explanation and reproduction :smile: 

Anyways, it seems like the animations code is run after the whole change detection runs which sets and re-sets the value as you showed. It is unsurprising since the animation code needs to kick off as an effect of change detection. And by how things are implemented it to me looks reasonable enough that that such very fast operations could be missed by the animations code.

___

One thing that I'd like to mention is that with your `setTimeout`s you are creating a macro events, which populate the queue and that's how you create a queue of relatively slow macro tasks which make this work, if instead of `setTimeout` you were to use promises you could see that the issue still remains (see [here](https://stackblitz.com/edit/angular-ivy-ts-strict-enabled-j7dw9c?file=src/app/app.component.ts)).
So to me it looks like you are sort of creating a queue of separate events for this scenario which is all squashed into/being treated as a single event by Angular, asking Angular to do things differently to me sounds like it would need to introduce unwanted (significant) delays :thinking: 

___

But regardless on the above, the most important thing that I'd like to ask you is: since the transition between the two states is so small that the animation code doesn't see it, what would be a real world scenario in which you'd actually need to detect that change? (since animations in the order of tens of milliseconds won't make much sense)

Is this all regarding callbacks? because if that's the case I would say that this is the correct behaviour (since no animation is being applied anyways since the state change is not detected) and that there are countless much better alternatives (like via setters) to detect when the values have changed.

**Comment by 4javier** [comment]
Hi @dario-piotrowicz, and thanks for taking the time to read the ticket.
When I opened it I was already almost sure it currently behaves as it's supposed to do.
You're right with saying with setTimeout I'm explicitly scheduling the router activation callback on the MacroTask queue.
I suspect it doesn't work the same with MicroTask because animation triggers callbacks could be scheduled on Macro one (but it's just a supposition of mine), so executed after both promise calls resolved, and reset property value to original state, like the sync callbacks do.
The reason why I'm doing that on purpose, is to try to realize a real sequencial routing transition animation.
Not one where you play around syncing the duration of leaving one with a delay of the entering one, but a setup where the entering component creation will be rendered animated only after previous component leaving animation has terminated.
No matter what their durations are.
I think that a router callback called asynchronously could be helpful.
But I understand perfectly how anything related to a basic task like routing has to be scheduled on stack to have highest priority on everything.
Maybe having another event structured the same as the current one but passing `isAsync` to its EventEmitter could solve the issue.
But I'm not too sure about that, because I'm still testing things to write an article about that.
And in any case, probably nothing worth too much efforts.

**Comment by dario-piotrowicz** [comment]
Ah I see, thanks so much for the clarification

Now I understand why you'd want this to work.

Just an idea, but could you implement your own routerOutlet alternative using [RouterOutletContract](https://angular.io/api/router/RouterOutletContract) (and/or by directly extend the `routerOutlet`)?
It would give you the leverage to implements the events they way you want and through it you could have more fine-grained control over the routing (as for example generating queues for the animation states)

**Comment by 4javier** [comment]
I was thinking about that or a directive to be added to existing `<router-outlet>`.
But I got to reach a right balance between "hack" and reliability first.
At the moment I'm tweaking a mix of promises and async observables that should let me managing both pre and post painting after routing.

**Comment by 4javier** [comment]
@dario-piotrowicz 
Sorry for pinging. I'm almost there with my project.
I managed to do what I want with animation registered as trigger.
Now I need to turn that in a customizable one, so I'm accessing `AnimationBuilder` API directly.
Everything works except for a problem in timings of router deactivate callback.
With declarative approach, I set wanted animation state as `(deactivate)` output bound expression.
Animation kicks in correctly.
With programmatically approach I listen to `ChildrenOutletContexts.context.deactivate` in code and run the `AnimationPlayer` inside subscriber.
Unexpectedly, animation starts after DOM has already been repainted, so target element is not present anymore, and obviously nothing gets animated.
I thought `AnimationPlayer.play()` itself was executed in a microtask, instead it looks like the magic happens inside the engine for transition animations.
Source code is quite complex, do you know where's the gist?
Here a repro of two different behaviours
https://stackblitz.com/edit/angular-ivy-ts-strict-enabled-9soeyf

**Comment by dario-piotrowicz** [comment]
Hi @4javier , no need to apologise, feel free to ping me whenever, I'm glad if I can help :slightly_smiling_face: 

___

Anyways, you are slightly off in your understanding of the issue, it is not that the query fails because it is executed too late, but it is the exact opposite, the query is executed too early!

When you use the `AnimationBuilder` API, the query call gets executed when you create the player, and not when you play it, so in your case you are querying for the red component at the very beginning when it is nowhere to be found.

You can see what I mean if you remove the `{optional: true}` in the query call, in that case you get the error in the console when the component is initialized and not when you play the animation:

![Peek 2022-09-18 16-34](https://user-images.githubusercontent.com/61631103/190915322-ca4c6e27-c6b8-4d71-9a4f-ac2f7bd4ee96.gif)

So to fix this issue you need to create the player on-demand and not immediately in the `ngAfterViewInit` (and the declarative approach does exactly this, generates new players every time).

So to get the query to work you can simply do:
```diff
  ngAfterViewInit() {
-    this.createPlayer();

    this.context.getContext('secondary')?.outlet?.deactivateEvents?.subscribe(
-      () => this.secondaryAnimationPlayer.play());
+      () => (this.createPlayer(), this.secondaryAnimationPlayer.play()));
  }
```

This should be it.... but things of course are never that simple :sweat_smile: 

___

There is another issue unfortunately (much more annoying!)

The declarative API prevents the routing to change (or more precisely, the element removal), but manually built players do not... so you also need to block the thing somehow :cry: 

I'll try to think if there is something clever that can be done there :cry: , but for now, for example I made your repro working by creating a no-op animation which is simply used to prevent the routing/element removal:
```ts
export const noopExitTrigger = 
  trigger('noopExitTrigger', [
    transition('rendered => cleared', [
      query('router-outlet + *', [
        animate(
          '1000ms',
          keyframes([
            style({ }),
          ])
        ),
      ], {optional: true}),
    ]),
  ]);
```

I know that it is super ugly and probably not what you want :cry: (but it at least does show that everything works fine if the element is kept in the DOM), you can see the result in this [fork of your repro](https://stackblitz.com/edit/angular-ivy-ts-strict-enabled-787r7a)

___

Again I will try to think if there is a nice solution there, but this seems a bit tricky :sweat: , but I figured you'd appreciate me sharing what I've got thus far :slightly_smiling_face: 

If you have any idea on how to "prevent the routing" please let me know :slightly_smiling_face:

**Comment by 4javier** [comment]
LOL.
In my real project I was actually defining `createAnimator(): AnimationPlayer` and calling `createAnimator().play()` on subscription, and in the repro I thought it was more straightforward to use a prebuilt instance, ignoring how it actually works.
(I thought it was its `init()` called by `play()` in charge of querying the element to be animated).
I got to admit AnimationBuilder is not an easy game, a bit for lacking docs, and a bit for a codebase quite obscure. But the latter is just me.
I was considering a different approach: since it looks like one step after the other I'm gonna reproduce declarative registration flow, I'd try to programmatically call it directly, maybe injecting `AnimationRendererFactory` or something similar.
After all, the only reason I don't want to use `@Component''s `animation` property is just because I want trigger name to be customizable as template attribute.
Do you think this path worths a try?

**Comment by dario-piotrowicz** [comment]
Ah yeah I see, it's an understandable mistake :slightly_smiling_face: , and yes there is barely any documentation regarding that, I created an issue for that a while back (https://github.com/angular/angular/issues/44588) but it didn't get much traction :sweat: 

About what you suggested regarding using something like `AnimationRendererFactory` I must admit that I am not too familiar with that, but I __think__ that that is not even exposed and that it could be quite difficult.
I __think__ writing your own [AnimationDriver](https://angular.io/api/animations/browser/AnimationDriver) could help, but again I cannot say I am 100% confident in it (and again it isn't going to be easy).

If possible I would try to say away from trying to reproduce things that are already there as that may be quite complex and there  may be many scenarios to consider.

But I do not want to discourage you with that, it may be worth to explore those things to get a feeling if things would be possible.

Anyways I am not still 100% sure on what you're implementing, but I would try to rely on stuff that's already there to try to get what you need. For example, if you don't want trigger names to be registered in the class decorator as you mentioned I have 2 ideas in mind that maybe could help with that.

One would be to create a custom trigger which is the only one that needs to be registered, then you could apply that trigger and combine it with params (and/or attributes in the element) to keep the target element in the DOM and to run manually your programmatic animation via the start callback (it's just an idea, I am sure there will be a bunch of problems to be ironed out to make it work), for example I am thinking of something like:
```
<div [@customTrigger]="{
 value: '', // some value here
 params: {
   dynamicTriggerName: 'myActualTriggerName', // use this one to retrieve the actual trigger and play it programaticaly
   targetElementQuery: 'route-routlet + *', // the query to specify the target element
   duration: '1s', // the duration to use in the customTrigger to keep the target element in the dom (using a no-op animation) 
 },
}"></div>
```

The second idea would be to keep everything as you have it now and delay the routing via a `canActivate` guard, you could apply the guard to all pages, check if a page contains an element with a special attribute or something like that, take the duration from that element (again, likely defined using an attribute) and use that to delay the routing.

Anyways this is just me thinking out loud, I am not really sure what the best solution would be here :sweat: 

What do you think? Does either idea sound like it could be a viable solution?

**Comment by dario-piotrowicz** [comment]
Hi @4javier sorry for not having replied in a while, I've been occupied with other stuff and also I wasn't sure what to make of this.

I've thought/tried of different possible solutions, like using router guards or manipulating the dom to keep the elements in the page (sort of the same as the animations code does). But without much luck (the closes I got was with a `canActivate` route guard, that can indeed delay the routing, but it can be used only if the route goes somewhere).

So I don't know how you could tackle this, the only thing that comes to mind is that Angular itself should provide a way to delay route changes...

Have you had any luck?

**Comment by 4javier** [comment]
No worries for late replies.
I'm convinced with current implementation there's no clean (or perhaps no at all) way to achieve my goal.
So I tried puzzling around with framework codebase.

My first attempt was to add a new `@Output` to `RouterOutlet`, emitting just before the component gets destroyed inside its `deactivate` routine. I then hooked to that my `AnimationPlayer` creation&play.
No joy for me even that way.
The problem is the same, so I think at some point in building/factoring/playing of the animation, something is put in MacroTask or alike, firing up when the element has already been cleared from DOM.

Now I'm trying to understand how the flow to `AnimationRenderer` creation works, always trying to get what makes declarative approach capable of kicking in before element gets nuked.
I'm focusing on this conditional branch that looks like the one matching the programmatical approach due to its `!type.data['animation']` condition
https://github.com/angular/angular/blob/main/packages/platform-browser/animations/src/animation_renderer.ts#L50

**Comment by dario-piotrowicz** [comment]
I see :slightly_smiling_face: 

Yeah I agree that a clean solution doesn't seem to be really possible :cry: 

One thing I think you should be careful about is that the routing either needs to be prevented or you need to keep the dom elements in the dom by yourself somehow. The dom elements relative to the declared triggers are kept in the dom and I _think_ that that is done on a relatively low level...
Anyways what I'm trying to say is that I would keep in mind the routing issue, because you could end up getting the animation kicked off before the routing event, but then your dom elements would get removed so you could end up on square one. So my gut feelings tell me that you either need to make sure programmatically animating elements are kept in the dom (which feels a bit tough to me) or that the routing gets delayed according to your animations (this seems more feasible to me but I didn't find a way to do it reliable/generally).

Anyways best of luck! I hope you can get to the bottom of it! :smiley: (keep us posted! :grinning:)
(if there'll be stuff you think I could help with let me know :slightly_smiling_face:)

**Comment by 4javier** [comment]
Hi @dario-piotrowicz 
Unfortunately my try with custom `AnimationBuilder` failed.
I managed to inject it and registering a trigger programmatically just to find out that's not that one doing the magic.
Declarative flow bypasses `AnimationBuilder` and calls `AnimationRendererFactory.createRenderer` directly from `ComponentFactory`.
I'm a bit short of ideas, now.
The only thing I can think of is a custom `AnimationRendererFactory` to be injected in my dynamic component.
It should ignore the `AnimationMetadata` passed to it by `ComponentFactory`'s call `ComponentDef` parameter, and read one injected by the structural directive creating the dynamic component.
But I'm quite tired and not too much clear headed right now, so if it sounds a dumb idea, let me now.

**Comment by 4javier** [comment]
@dario-piotrowicz 
Sorry again. I don't know where else asking about animation internals.
I wrote an alternative AnimationRenderer that gets injected in a component wrapping the children I want to animate.
It works as expected for `¢reateRenderer`, but I noticed that its `begin()` and `end()` method implementations gets ignored by change detection.
I debugged a little and noticed that while `ComponentFactory.create()` correctly looks for renderer token by injector getter
```ts
const rendererFactory = rootViewInjector.get(RendererFactory2, null);
```
change detection mechanism retrieves that accessing directly `lView`
```ts
const rendererFactory = lView[RENDERER_FACTORY];
```
Is that how it's supposed to act?
Does that mean that change detection can take in account only root injected renderer factory?

**Comment by dario-piotrowicz** [comment]
Hi @4javier no problem at all :slightly_smiling_face: 

I am so sorry but I have no idea :sweat: , I would suggest you to maybe create an issue mentioning that it is a potential bug?

____

My guess (totally on the top of my head, I might be completely wrong! :sweat_smile:) is that the code:
```ts
const rendererFactory = lView[RENDERER_FACTORY]
```
is correct as you don't have access to the correct injector and/or need to get the renderer factory based on the views structure...

I've had a look around and it seems like this:
https://github.com/angular/angular/blob/7ab10a503132772e3b57acd440e29073b2f2731d/packages/core/src/render3/instructions/shared.ts#L138
is the only place in which the field is set, so maybe you could check out which `createLView()` is creating your component views and check there how it gets the renderer factory and if that there it should take it from the injector?

By looking at the code it looks like it should take that from the injector?
https://github.com/angular/angular/blob/917816f6725cae8a4e6bf03f104dea00d79f822c/packages/core/src/render3/component_ref.ts#L155

Maybe that's what you need to check? that the wrong rendered factory seems to end up in your `lview`?
What do you think?

By the way, sorry I could not be of much help, if you want you can share your code (or part of it) and I could try to debug it and see if I can help you in that way?

**Comment by 4javier** [comment]
Hi @dario-piotrowicz 
Sorry for late reply, that's due to the same reason I still didn't post any code for this project:
I keep rewriting it from scratch. I'm at least at the 5-6 total redesign, started as a router-outlet replacement, then a wrapping directive, now an alternative AnimationRendererFactory plugin.
It's all so messed up I didn't even commit any change in weeks.
I thought I was at a good point but then realized that my system could have (maybe, not even sure...) worked only when routed components had their enter/exit animations defined on their host element.
I think most of the times people put their triggers on the first node element of their component.
Nail to the coffin was noticing how the system swallows up exiting animations defined on node elements.
I suspect the culprit could be the call to `triggerLeaveAnimation(element, context, destroyAfterComplete, defaultToFallback)`, executed with `destroyAfterComplete=true` for host nodes
https://github.com/angular/angular/blob/d07241dd12c2d4dd16cb316711fc60221d80c3a7/packages/animations/browser/src/render/transition_animation_engine.ts#L410
while called with the same argument with `false` value for child nodes
https://github.com/angular/angular/blob/d07241dd12c2d4dd16cb316711fc60221d80c3a7/packages/animations/browser/src/render/transition_animation_engine.ts#L328
in the latter instructing `AnimationTransitionNamespace` to not add their leaving animations to the `onDone` callback of `optimizeGroupPlayer`, thus killing them.
https://github.com/angular/angular/blob/d07241dd12c2d4dd16cb316711fc60221d80c3a7/packages/animations/browser/src/render/transition_animation_engine.ts#L359-L368

I got the feeling that choice has been made on purpose to avoid concurrent rendering of entering and leaving animations.
But maybe I totally misinterpreted rendering flow.

Thanks for your time and your patience.
Wish you the best working on this really strange piece of code. 
:beers:

**Comment by dario-piotrowicz** [comment]
@4javier I am so sorry, I forgot to look into this in the past 2 weeks, I'll be a bit busy for the next few days, I'll check things out in a week or so, so sorry about this 🙏 

Also if you have any new developments please let me know, could you also clarify what is now the specific thing you're tring to do?

**Comment by 4javier** [comment]
Hi Dario.
Absolutely no apologies needed.
You've been kind and helpful beyond any expectation. :smile:
I was just thanking you for your efforts in answering. 
I can't remember exactly what my last try was, because I gave up on the project, considering that it seems current state of animation framework is not expected to work with real sequential namespace triggering.
I'm using all I learned in this time to write a couple blog post just to not waste many interesting info I got, but I not intend to keep going with the project.
Thanks again and have a nice holiday and happy new year. :tada:

**Comment by dario-piotrowicz** [comment]
Ah ok I see, I'll leave it be then 🙂 

you're very welcome, although I feel like I didn't manage to help you much 😓  (you're too kind thanks! 😊)

I'm sorry you had to give up on the project, but it does feel like things are a bit tough to handle with the current animations implementation 😢 

When they're ready please share the blog posts, I'd love to read them! 😄 
(in any case as usual if there is anything I can help with or you decide to give the project another try just let me know 🙂)

Have a nice holiday and a happy new year too! 🎉 🍾

**Comment by 4javier** [comment]
> When they're ready please share the blog posts, I'd love to read them!

I don't think they can be such interesting for you, because I touched just initial part of animation registering.
But since you're asking:
[Here they are](https://dev.to/gianpiero_errigo/series/21120)
Feel free to get in touch in cooments for anything. :slightly_smiling_face:

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
